### PR TITLE
Add embedder_config in core data source

### PR DESCRIPTION
**Pull request by flvndvd**
## Description

This PR introduces the `embedder_config` field in the core data source, laying the groundwork to support multiple embedders. This preliminary step is essential before upgrading our current embedder to the latest version. Although this new field is not yet used in the core, the immediate goal is to begin saving it when creating new data sources so that we can backfill existing records.

The plan to fully transition to this structure involves:
1. Adding `embedder_config` and updating API calls.
2. Backfilling `embedder_config` and deprecating the legacy top-level embedder configuration.
3. Starting to use the new `embedder_config`.
4. Handling the `shadow_embedder`.

Tested locally:
```
{
  "provider_id": "openai",
  "model_id": "text-embedding-ada-002",
  "splitter_id": "base_v0",
  "embedder_config": {
    "embedder": {
      "provider_id": "openai",
      "model_id": "text-embedding-ada-002",
      "splitter_id": "base_v0"
    },
    "shadow_embedder": null
  },
  "extras": null,
  "max_chunk_size": 512,
  "qdrant_config": { "cluster": "cluster-0", "shadow_write_cluster": null }
}
```

## Risk

Low risk.

## Deploy Plan

⚠️ We first needs to deploy `core` before deploying `front`.

### CODE REVIEW

File: core/src/data_sources/data_source.rs [hunk 1]
```diff
@@ -402,12 +402,30 @@ pub struct DocumentVersion {
...
     pub provider_id: ProviderID,
     pub model_id: String,
-    pub extras: Option<Value>,
     pub splitter_id: SplitterID,
+
+    // Optional until we update api calls to pass this body.
+    pub embedder_config: Option<EmbedderDataSourceConfig>,
+
+    pub extras: Option<Value>,
     pub max_chunk_size: usize,
```
Comment (spolu) [inline core/src/data_sources/data_source.rs:429]: We def want to move max_chunk_size as well as bumping chunk_size will require similar re-embedding :+1:

**Review by spolu — APPROVED** [review]
Let's add max_chnk_size to embedder_config and feel free to merge once done :+1:

**Comment by spolu** [comment]
Do we use extras anywhere? If not maybe nuke?

**Comment by spolu** [comment]
Confirmed we don't. Maybe worth removing (even if that's likely the next PR)?

File: types/src/core/data_source.ts [hunk 2]
```diff
@@ -1,12 +1,26 @@
...
 export type CoreAPIDataSourceConfig = {
+  // TODO(2024-05-27 flav) Remove once migrated to embedder.
   provider_id: string;
   model_id: string;
+  splitter_id: string;
+
+  embedder_config: EmbedderConfigType;
+
   // eslint-disable-next-line @typescript-eslint/no-explicit-any
   extras?: any | null;
```
Comment (spolu) [inline types/src/core/data_source.ts:23]: We likely want to remove it from here in that PR?

File: core/src/data_sources/data_source.rs (hunk 1 above)
Comment (flvndvd) [reply core/src/data_sources/data_source.rs:429]: 🤔 We chat live about this on Friday and you said no. Anyway, I'll update.

**Review by spolu — COMMENTED** [review]
Should this PR include an update to postgres.rs ?

File: types/src/core/data_source.ts (hunk 2 above)
Comment (flvndvd) [reply types/src/core/data_source.ts]: Done.

File: core/src/data_sources/data_source.rs (hunk 1 above)
Comment (spolu) [reply core/src/data_sources/data_source.rs:429]: I was wrong I think :)

**Comment by flvndvd** [comment]
> Should this PR include an update to postgres.rs ?

Given that we don't add a field and only rely on the existing `config_json` field, only the API needs to be updated.

**Comment by spolu** [comment]
> Given that we don't add a field and only rely on the existing config_json field, only the API needs to be updated. Am I missing something here?

True! Forgot it was one json text :)
That being said, thought you wanted to move to a JSON type?

**Pull request merged**